app/src/components/GameRoutes.tsx

```diff
@@ -15,23 +27,88 @@ function CurrentPlayerProvider(props: {
   joinCode: string
   children: React.ReactNode
 }) {
-  const { data, loading } = useCurrentPlayerQuery({
+  const currentAuth = React.useContext(CurrentAuthContext)
+  const [player, setPlayer] = React.useState<any>([])
+  const [attemptedGameLoad, setAttemptedGameLoad] = React.useState(false)
+  const [isLoading, setLoading] = React.useState(true)
+  const [joinGame] = useJoinGameMutation()
+
+  const [getPlayers] = useCurrentPlayerLazyQuery({
     variables: {
       joinCode: props.joinCode,
       clientUuid: clientUuid()
+    },
+    onCompleted: data => {
+      setPlayer(data?.players[0])
+      setLoading(false)
+    }
+  })
+
+  // This "skip logic" is sort of hacky but handles a bug in useQuery
+  // https://github.com/apollographql/react-apollo/issues/3505#issuecomment-568112061
+  const [skipCallback, setSkipCallback] = React.useState(false)
+  const [loadGame] = useLazyQuery(GameByJoinCodeDocument, {
+    variables: { joinCode: props.joinCode?.toLocaleUpperCase() },
+    onCompleted: async (data: GameByJoinCodeQuery) => {
+      if (skipCallback) {
+        return
+      }
+
+      if (data && data.games[0]) {
+        // If we found a game from the game code, now try to join it
+        const registration = await joinGame({
+          variables: {
+            gameId: data.games[0].id,
+            clientUuid: clientUuid()
+          }
+        })
+        if (registration.data?.joinGame) {
+          await currentAuth.setJwtToken(registration.data.joinGame.jwt_token)
+        }
+
+        // If we successfully joined a game and set the JWT token, re-fetch the
+        //  current player (which should now be set)
+        getPlayers()
+      } else {
+        setLoading(false)
+      }
+
+      setSkipCallback(true)
+      setAttemptedGameLoad(true)
     }
   })
 
-  if (!loading && !data?.players[0]) {
+  // To start, fetch the current player and the games they are a part of
+  React.useEffect(() => {
```

Comment:
this replaces the initial ` const { data, loading } = useCurrentPlayerQuery({` query